core/record/_record/recordReader.js

```diff
@@ -107,6 +107,38 @@ export const visitDescendantsAndSelf = (node, visitor) => (record) => {
   }
 }
 
+/**
+ * Finds a the parent node of the specified node def, starting from the specified parent node and traversing
+ * the single entities, if any, down to the correct parent node.
+ */
+export const getNodeParentInDescendantSingleEntities =
+  ({ survey, parentNode, nodeDefUuid }) =>
+  (record) => {
+    const nodeDefParent = SurveyNodeDefs.getNodeDefByUuid(Node.getNodeDefUuid(parentNode))(survey)
+    const nodeDef = SurveyNodeDefs.getNodeDefByUuid(nodeDefUuid)(survey)
+    if (!NodeDef.isDescendantOf(nodeDefParent)(nodeDef)) {
+      throw new Error('target node is not a descendant of the specified parent entity')
+    }
+    const nodeDefHierarchy = NodeDef.getMetaHierarchy(nodeDef)
+    const nodeDefParentHierarchy = NodeDef.getMetaHierarchy(nodeDefParent)
+    const hierarchyToVisit =
+      nodeDefHierarchy.length > nodeDefParentHierarchy.length + 1
+        ? nodeDefHierarchy.slice(nodeDefParentHierarchy.length + 1)
+        : []
+    return hierarchyToVisit.reduce((parentNodeCurrent, descendantDefUuid) => {
+      const nodeDefDescendant = SurveyNodeDefs.getNodeDefByUuid(descendantDefUuid)(survey)
+      if (NodeDef.isSingleEntity(nodeDefDescendant)) {
+        return getNodeChildByDefUuid(parentNodeCurrent, descendantDefUuid)(record)
+      } else {
+        throw new Error(
+          `the target node ${NodeDef.getName(nodeDef)} is inside a multiple entity: ${NodeDef.getName(
+            nodeDefDescendant
+          )}`
+        )
+      }
+    }, parentNode)
```

Comment:
can be simple to understand if you move the reduce above of the function return

Reply:
maybe I can use a forEach instead... the reduce itself is not so clear in this case